File: src/v2/shared/platform/tauri/feature-ports/managedAuth.ts
import { invoke } from "@tauri-apps/api/core";

import {
  assertManagedAuthAccountMutation,
  assertManagedAuthConnectionActionRequest,
  assertManagedAuthLoginMethod,
  assertManagedAuthRemovalMutation,
  assertManagedAuthSessionId,
  assertStartManagedAuthLoginRequest,
  parseManagedAuthLoginSession,
  parseManagedAuthMutationResult,
  parseManagedAuthOverview,
  parseManagedAuthRemovalPreview,
  type ManagedAuthPort,
} from "../../../features/managed-auth";

export function createManagedAuthPort(): ManagedAuthPort {
  return {
    getOverview: async () =>
      parseManagedAuthOverview(
        await invoke<unknown>("managed_auth_get_overview"),
      ),
    startLogin: async (request) =>
      parseManagedAuthLoginSession(
        await invoke<unknown>("managed_auth_start_login", {
          request: assertStartManagedAuthLoginRequest(request),
        }),
      ),
    getLoginSession: async (sessionId) =>
      parseManagedAuthLoginSession(
        await invoke<unknown>("managed_auth_get_login_session", {
          sessionId: assertManagedAuthSessionId(sessionId),
        }),
      ),
    cancelLogin: async (sessionId) =>
      parseManagedAuthLoginSession(
        await invoke<unknown>("managed_auth_cancel_login", {
          sessionId: assertManagedAuthSessionId(sessionId),
        }),
      ),
    reopenLogin: async (sessionId) =>
      parseManagedAuthLoginSession(
        await invoke<unknown>("managed_auth_reopen_login", {
          sessionId: assertManagedAuthSessionId(sessionId),
        }),
      ),
    switchLoginMethod: async (sessionId, method) =>
      parseManagedAuthLoginSession(
        await invoke<unknown>("managed_auth_switch_login_method", {
          sessionId: assertManagedAuthSessionId(sessionId),
          method: assertManagedAuthLoginMethod(method),
        }),
      ),
    setDefaultAccount: async (accountId, expectedRevision) =>
      parseManagedAuthMutationResult(
        await invoke<unknown>("managed_auth_set_default_account", {
          request: assertManagedAuthAccountMutation(
            accountId,
            expectedRevision,
          ),
        }),
      ),
    previewAccountRemoval: async (accountId, expectedRevision) =>
      parseManagedAuthRemovalPreview(
        await invoke<unknown>("managed_auth_preview_account_removal", {
          request: assertManagedAuthAccountMutation(
            accountId,
            expectedRevision,
          ),
        }),
      ),
    removeAccount: async (previewId, accountId, expectedRevision) =>
      parseManagedAuthMutationResult(
        await invoke<unknown>("managed_auth_remove_account", {
          request: assertManagedAuthRemovalMutation(
            previewId,
            accountId,
            expectedRevision,
          ),
        }),
      ),
    applyConnectionAction: async (request) =>
      parseManagedAuthMutationResult(
        await invoke<unknown>("managed_auth_apply_connection_action", {
          request: assertManagedAuthConnectionActionRequest(request),
        }),
      ),
  };
}

File: src-tauri/src/commands/managed_auth.rs
//! IPC facade for the unified managed-auth control plane.

use std::sync::Arc;

use tauri::State;

use crate::services::managed_auth::{
    validate_session_id, ManagedAuthAccountMutationRequest, ManagedAuthAccountRemovalPreview,
    ManagedAuthAccountRemovalRequest, ManagedAuthConnectionActionRequest, ManagedAuthErrorDto,
    ManagedAuthLoginMethod, ManagedAuthLoginSessionSnapshot, ManagedAuthMutationResult,
    ManagedAuthOverview, NativeManagedAuthService, StartManagedAuthLoginRequest,
};

pub struct ManagedAuthState(pub(crate) Arc<NativeManagedAuthService>);

#[tauri::command]
pub fn managed_auth_get_overview(state: State<'_, ManagedAuthState>) -> ManagedAuthOverview {
    state.0.overview()
}

#[tauri::command(rename_all = "camelCase")]
pub fn managed_auth_start_login(
    request: StartManagedAuthLoginRequest,
    state: State<'_, ManagedAuthState>,
) -> Result<ManagedAuthLoginSessionSnapshot, ManagedAuthErrorDto> {
    request.validate()?;
    state.0.start_login(request)
}

#[tauri::command(rename_all = "camelCase")]
pub fn managed_auth_get_login_session(
    session_id: String,
    state: State<'_, ManagedAuthState>,
) -> Result<ManagedAuthLoginSessionSnapshot, ManagedAuthErrorDto> {
    validate_session_id(&session_id)?;
    state.0.get_login_session(&session_id)
}

#[tauri::command(rename_all = "camelCase")]
pub fn managed_auth_cancel_login(
    session_id: String,
    state: State<'_, ManagedAuthState>,
) -> Result<ManagedAuthLoginSessionSnapshot, ManagedAuthErrorDto> {
    validate_session_id(&session_id)?;
    state.0.cancel_login(&session_id)
}

#[tauri::command(rename_all = "camelCase")]
pub fn managed_auth_reopen_login(
    session_id: String,
    state: State<'_, ManagedAuthState>,
) -> Result<ManagedAuthLoginSessionSnapshot, ManagedAuthErrorDto> {
    validate_session_id(&session_id)?;
    state.0.reopen_login(&session_id)
}

#[tauri::command(rename_all = "camelCase")]
pub fn managed_auth_switch_login_method(
    session_id: String,
    method: ManagedAuthLoginMethod,
    state: State<'_, ManagedAuthState>,
) -> Result<ManagedAuthLoginSessionSnapshot, ManagedAuthErrorDto> {
    validate_session_id(&session_id)?;
    state.0.switch_login_method(&session_id, method)
}

#[tauri::command(rename_all = "camelCase")]
pub fn managed_auth_set_default_account(
    request: ManagedAuthAccountMutationRequest,
    state: State<'_, ManagedAuthState>,
) -> Result<ManagedAuthMutationResult, ManagedAuthErrorDto> {
    request.validate()?;
    state
        .0
        .set_default_account(&request.account_id, &request.expected_revision)
        .map_err(ManagedAuthErrorDto::from_core)
}

#[tauri::command(rename_all = "camelCase")]
pub fn managed_auth_preview_account_removal(
    request: ManagedAuthAccountMutationRequest,
    state: State<'_, ManagedAuthState>,
) -> Result<ManagedAuthAccountRemovalPreview, ManagedAuthErrorDto> {
    request.validate()?;
    state
        .0
        .preview_account_removal(&request.account_id, &request.expected_revision)
        .map_err(ManagedAuthErrorDto::from_core)
}

#[tauri::command(rename_all = "camelCase")]
pub fn managed_auth_remove_account(
    request: ManagedAuthAccountRemovalRequest,
    state: State<'_, ManagedAuthState>,
) -> Result<ManagedAuthMutationResult, ManagedAuthErrorDto> {
    request.validate()?;
    state
        .0
        .remove_account(
            &request.preview_id,
            &request.account_id,
            &request.expected_revision,
        )
        .map_err(ManagedAuthErrorDto::from_core)
}

#[tauri::command(rename_all = "camelCase")]
pub fn managed_auth_apply_connection_action(
    request: ManagedAuthConnectionActionRequest,
    state: State<'_, ManagedAuthState>,
) -> Result<ManagedAuthMutationResult, ManagedAuthErrorDto> {
    request.validate()?;
    state.0.apply_connection_action(&request)
}
